articles/introduction-to-evolutionary-algorithms/index.md

```diff
@@ -0,0 +1,404 @@
+        child[i] = b[i] if i > mid_point else a[i]
+    return ''.join(child)
+
+
+def mutate(a):
+    a = list(a)
+    for i in range(len(a)):
+        if random.random() < mutationRate:
+            a[i] = random.choice(string.ascii_letters)
+    return ''.join(a)
+
+
+def generation(mating_pool, number_of_gen):
+    for i in range(len(population)):
+        a = random.choice(mating_pool)
+        b = random.choice(mating_pool)
+        child = mutate(crossover(a, b))
+        population[i] = child
+    number_of_gen += 1
+    return number_of_gen
+
+
+def evaluate():
+    score = 0
+    best = 0
+    for i, val in enumerate(population):
+        c = sum([i == j for i, j in zip(target, val)])
+        if c > score:
+            best = i
+            score = c
+    return best, score
+
+
+target = "Hello"
+mutationRate = 0.1
+n = 50
+population = [''.join(random.choices(string.ascii_letters, k=len(target))) for _ in range(n)]
+print("The time now: ", time.asctime(time.localtime(time.time())))
+initTime = time.time()
+print("The population: ")
+print(population)
+print("\n Generations: ")
+generations = 0
+bestS, bestV = 0, ""
+while generations < 1000:
+    f_scores = [fitness(val, target) for val in population]
+    pool = selection(population)
+    generations = generation(pool, generations)
+    bestVal, bestScore = evaluate()
+    print("Gen: ", generations, population[bestVal], bestScore)
+    if bestS < bestScore:
+        bestV = population[bestVal]
+        bestS = bestScore
+    if bestScore == len(target):
+        break
+print("\n Best Match: ", bestV, bestS)
+print("The time now: ", time.asctime(time.localtime(time.time())))
+print("Total time taken: ", time.time() - initTime)
+
+````
+The output:
+````
+The time now:  Mon May 10 09:32:25 2021
+The population:
+['tlRWu', 'kBDkM', 'YxePb', 'TyYsM', 'UpmBu', 'yInSD', 'rOfLM', 'yLovB', 'vtMBA', 'LhVUL', 'TPFQv', 'xrBOZ', 'qHMdC', 'YSHlV', 'uyOLR', 'rTczE', 'aokjT', 'QvJvS', 'vffVO', 'eSnFy', 'IaBmO', 'zXkzL', 'IlYzI', 'DMEKO', 'sSFGW', 'aWlAm', 'QJFTH', 'ruKJl', 'SfGPJ', 'huOOi', 'OfNdu', 'UpqKP', 'JIYcg', 'UJorD', 'MJZro', 'NGGcU', 'MQQKs', 'twSRo', 'EueKl', 'umvDz', 'iDNrM', 'MjbON', 'Hzyht', 'Onrls', 'bUgSl', 'bhZPk', 'fnSXd', 'xvgmz', 'AvjNj', 'etFMZ']
+ Generations:
+Gen:  1 HzZro 2
+Gen:  2 Hzylo 3
+Gen:  3 Hzrlo 3
+Gen:  4 HJrlo 3
+Gen:  5 HAHlo 3
+Gen:  6 Hzylo 3
+Gen:  7 HeHlo 4
+Gen:  8 HeHlo 4
+Gen:  9 HeUlo 4
+Gen:  10 HeDlo 4
+Gen:  11 HeHlo 4
+Gen:  12 HeUlo 4
+Gen:  13 HeUlo 4
+Gen:  14 HeHlo 4
+Gen:  15 HeBlo 4
+Gen:  16 HePlo 4
+Gen:  17 Heplo 4
+Gen:  18 Hemlo 4
+Gen:  19 HeKlo 4
+Gen:  20 HeHlo 4
+Gen:  21 HeYlo 4
+Gen:  22 HeRlo 4
+Gen:  23 HeYlo 4
+Gen:  24 Hello 5
+ Best Match:  Hello 5
+The time now:  Mon May 10 09:32:25 2021
+Total time taken:  0.0304110050201416
+````
+You can see that the algorithm took only 0.03 s to compute the output.
+We have successfully dissected and coded an EA. Our monkey finally learned to say 'Hello'.
+Yay! 🥳
+
```

Comment:
Remember the link I had asked for to access the full code on replit? Please add it here so that readers can test the code for themselves on replit.